hudi-common/src/main/java/org/apache/hudi/common/table/read/buffer/FileGroupRecordBuffer.java

```diff
@@ -56,27 +53,29 @@
 
 import java.io.IOException;
 import java.io.Serializable;
+import java.util.HashSet;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Locale;
 import java.util.Map;
+import java.util.Set;
 import java.util.function.Function;
 
 import static org.apache.hudi.common.config.HoodieCommonConfig.DISK_MAP_BITCASK_COMPRESSION_ENABLED;
 import static org.apache.hudi.common.config.HoodieCommonConfig.SPILLABLE_DISK_MAP_TYPE;
 import static org.apache.hudi.common.config.HoodieMemoryConfig.MAX_MEMORY_FOR_MERGE;
 import static org.apache.hudi.common.config.HoodieMemoryConfig.SPILLABLE_MAP_BASE_PATH;
-import static org.apache.hudi.common.model.HoodieRecordMerger.PAYLOAD_BASED_MERGE_STRATEGY_UUID;
 import static org.apache.hudi.common.table.log.block.HoodieLogBlock.HeaderMetadataType.INSTANT_TIME;
 
 abstract class FileGroupRecordBuffer<T> implements HoodieFileGroupRecordBuffer<T> {
+  protected final Set<String> usedKeys = new HashSet<>();
```

Comment:
There is a possibility of duplicate keys in a file and there is an expectation that updates are applied to both rows. See `Test only insert for source table in dup key without preCombineField` for an example. We need to figure out if there is a better way to handle this.

Reply:
The fix is not right, the discrepencies come from the log comparison, for old write merge handle, we do not remove the log record when a base record key matches it, and bookeep a set of log record keys that have been matched with base(used for skipping for those non matched log records).

While the FG reader current remove the log record directly, which makes it impossbie to match the same log records with duplicates in base multiple times.

Reply:
I'm wondering if this is a valid case, because the duplicates in base only occurs in pkless table, while in the test case, the table has a primary key but it sets up to allow duplicates for the first commit intentionally, seems incorrect. We should limit the "allow duplicates" for incoming records to just `INSERT` operation for pkless table, just like the doc says in `HoodieConcatHandle`.

Reply:
Fired a discussion here: https://github.com/apache/hudi/pull/6824#issuecomment-3182443235

Reply:
For the test case that fails without this, should we just update it to be PK-less?